### DISCUSSION TITLE
import(..) returns undefinded instead of results of resolution

### DISCUSSION THREAD

**Discussion by gerwinbrunner** [pkosiec/mongo-seeding#155] [Q&A] [answered]
When the `import(..)` function is called the results are not returned, but a `undefined`
This leads to problems when using trying to use this package with cypress.

The for loop does not return the results of the promises
https://github.com/pkosiec/mongo-seeding/blob/master/core/src/importer/index.ts#L25-L28

If we would use a map and return instead of the `for` loop the issue could be solved.
What do you think?

**Comment by gerwinbrunner** [comment]
This is the actual error message:

The task returned undefined. You must return a promise, a value, or null to indicate that the task was handled.

**Comment by gerwinbrunner** [comment]
So after some investigation I found the issue to be that this function does not return anything - so it returns undefined.

https://github.com/pkosiec/mongo-seeding/blob/master/core/src/index.ts#L119

So if we add a `return true` all is good :)

**Comment by pkosiec** [comment]
Hi @gerwinbrunner, thank you for your contribution!

Yes, you're right - currently, the import function resolves with the `undefined` value. However, I don't see any problem with this approach. I would say it's common to not return anything (which result in `undefined` in JS) when there is nothing to return. I wouldn't feel comfortable to return `true` value when there are two options:
- function returns `true`
- function throws an error

In this case `void` or `undefined` seems just right to me, because basically the result doesn't matter: there is success (no error) or failure (error).

Could you please elaborate a bit about the issues you encounter with using `mongo-seeding`  along with Cypress? Are you sure the problem really occurs in `mongo-seeding`, not in the way you use cypress? Please paste a snippet.

Thanks!

**Comment by gerwinbrunner** [comment]
Sure!

in cypress I define a task to seed the data:

```
module.exports = (on) => {
  on("task", {
    "seed:database:baseAccountsEnabled": () => {
      const config = {
        database: "mongodb://localhost:3051/meteor",
        dropDatabase: true,
      };
      const mongoSeeder = new SeederClass(config);
      // console.log("seed:database:baseAccountsEnabled start");
      const p = path.resolve(folder);
      // console.log("seed:database:baseAccountsEnabled path", p);

      const collections = mongoSeeder.readCollectionsFromPath(p, {
        extensions: ["js", "json", "ts"],
        //transformers: [Seeder.Transformers.replaceDocumentIdWithUnderscoreId],
      });
      return mongoSeeder.import(collections);
    },
  });
};

```

Calling this task raises the error mentioned in a previous post. The data import is anyway done  successfully before the error is raised.

The return of the `undefined` raises the error - this is documented in the cypress task definition here:
https://docs.cypress.io/api/commands/task.html#Command

**Comment by gerwinbrunner** [comment]
I found a workaround for now without having to forge you library.
Its kind of ugly, but works.

```
module.exports = (on) => {
  on("task", {
    "seed:database:baseAccountsEnabled": async () => {
      const config = {
        database: "mongodb://localhost:3051/meteor",
        dropDatabase: true,
      };
      const mongoSeeder = new SeederClass(config);
      // console.log("seed:database:baseAccountsEnabled start");
      const p = path.resolve(folder);
      // console.log("seed:database:baseAccountsEnabled path", p);

      const collections = mongoSeeder.readCollectionsFromPath(p, {
        extensions: ["js", "json", "ts"],
        //transformers: [Seeder.Transformers.replaceDocumentIdWithUnderscoreId],
      });

      const prom = mongoSeeder.import(collections);
      await prom;
      return true;
    },
  });
};


```

**Comment by pkosiec** [comment, marked as answer]
@gerwinbrunner Yup, the workaround looks like it is the best approach here. I would only slightly modify the two ending lines for the function:

```javascript
await mongoSeeder.import(collections);
return true;
```

It seems right to me - because this is specific requirement from Cypress. And as I wrote before, resolving promise with `undefined` is just as fine as returning from function with `undefined`.

Do you feel convinced that this is not the issue of Mongo Seeding? 😄 If so, can we close this issue and PR?

**Comment by gerwinbrunner** [comment]
Yeah, sure - the workaround is fine for me.

**Comment by pkosiec** [comment]
I wouldn't treat it as a workaround, but in this case as a proper solution 🙂 Anyway, thanks for the openness and good open source attitude 🙂 Cheers!
